In [44]:
import pandas as pd

df=pd.read_csv("sales_data.csv")
df.head()

order_id     customer_name   region           product   sales_rep  \
0  ELX-1000          TechMart    North   Fitbit Charge 6    Jane Doe   
1  ELX-1001          TechMart  Central        Power Bank  John Smith   
2  ELX-1002  Electronics Plus  Central        Power Bank    Jane Doe   
3  ELX-1003  Electronics Plus     West  Screen Protector  Bob Wilson   
4  ELX-1004        TechMart       East  Wireless Charger  Mike Davis   

   order_date  quantity  unit_price discount  shipping_cost    revenue  \
0  2023-04-25         9      887.72     0.17          22.84  $6,631.27   
1  2023-01-16         6      274.09     0.13           5.53    1430.75   
2  2023-08-18        38      344.66     0.22          20.18   10215.72   
3  2023-03-21        14     1149.30     0.08           6.85   14802.98   
4  2023-05-16         3      879.73     0.13          24.46     2296.1   

   total_cost priority          sales_rep_email  
0     6654.11   medium       jane.doe@yahoo.com  
1     1436.28     HIGH  john.smith@business.org  
2    10235.90     High       jane.doe@yahoo.com  
3    14809.83   medium     BOB.WILSON@YAHOO.COM  
4     2320.56   Medium                      NaN

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1050 entries, 0 to 1049
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   order_id         1050 non-null   object 
 1   customer_name    1050 non-null   object 
 2   region           1050 non-null   object 
 3   product          1050 non-null   object 
 4   sales_rep        1050 non-null   object 
 5   order_date       1050 non-null   object 
 6   quantity         1050 non-null   int64  
 7   unit_price       1050 non-null   float64
 8   discount         1029 non-null   object 
 9   shipping_cost    1050 non-null   float64
 10  revenue          1016 non-null   object 
 11  total_cost       1050 non-null   float64
 12  priority         918 non-null    object 
 13  sales_rep_email  899 non-null    object 
dtypes: float64(3), int64(1), object(10)
memory usage: 115.0+ KB


Looking at the DataFrame info, several columns have incorrect data types that should be converted:
Columns that need conversion:

*order_date (currently object)* - Should be converted to datetime64 since this represents dates  
*discount (currently object)* - Should be converted to float64 since discounts are typically numeric values (percentages or amounts)  
*revenue (currently object)* - Should be converted to float64 since revenue is a monetary/numeric value.

*customer_name*

In [46]:
df.customer_name.unique()

array(['TechMart', 'Electronics Plus', '  TechMart  ', 'GadgetHub',
       'Digital World', 'Smart Store', '  Smart Store  ',
       '  Digital World  ', '  GadgetHub  ', '  Electronics Plus  ',
       '    Digital World    ', '    Smart Store    ',
       '      Electronics Plus      ', '  Smart Store   Inc.',
       '  TechMart   Inc.', 'GadgetHub Inc.', 'Electronics Plus Inc.',
       'TechMart Inc.', 'Digital World Inc.', 'Smart Store Inc.',
       '  Electronics Plus   Inc.', '  GadgetHub   Inc.'], dtype=object)

In [47]:
import re

df['customer_name'] = (
    df['customer_name']
    # Remove leading and trailing spaces
    .str.strip() 

    # Replace multiple spaces with a single space
    .apply(lambda x: re.sub(r'\s+', ' ', x))  

    # Remove 'Inc.', 'Inc', and variations with trailing spaces
    .str.replace(r'\s*Inc\.*$', '', regex=True)
    
    # Convert to title case for consistency 
    .str.title()  
)

In [48]:
df.customer_name.unique()

array(['Techmart', 'Electronics Plus', 'Gadgethub', 'Digital World',
       'Smart Store'], dtype=object)

*region*

In [49]:
df.region.unique()

array(['North', 'Central', 'West', 'East', 'South', 'west', 'east',
       'central', 'NORTH', 'WEST', 'EAST', 'CENTRAL', 'north', 'south',
       'SOUTH'], dtype=object)

In [50]:
df['region']=df.region.str.capitalize()

In [51]:
df.region.unique()

array(['North', 'Central', 'West', 'East', 'South'], dtype=object)

*product*

In [52]:
df['product'].value_counts().sort_index()

product
AIRPODS PRO              2
AirPods Pro             50
AirPods-Pro              5
Apple AirPods Pro        2
Apple MacBook Pro        1
Apple Watch Series 9    62
Apple iPhone 15          5
Bluetooth Speaker       61
Dell XPS 13             62
Fitbit Charge 6         62
Galaxy S24               2
Gaming Headset          54
IPHONE 15                1
IPHONE CASE              2
Laptop Stand            68
MACBOOK PRO              2
MacBook Pro             43
MacBook-Pro              1
Phone Case               1
Phone Holder            59
Portable Charger         2
Power Bank              47
PowerBank                2
SAMSUNG GALAXY S24       3
SCREEN PROTECTOR         1
Samsung  Galaxy  S24     3
Samsung Galaxy S24      54
Screen Protector        46
Screen-Protector         3
Sony WH-1000XM5         55
Type-C Cable             3
USB C Cable              6
USB-C Cable             42
USBC Cable               1
WIRELESS CHARGER         3
Wireless Charger        78
Wireless-Charger    

*order_date*

In [53]:
# Convert order_date to datetime
df['order_date'] = pd.to_datetime(df['order_date'])

*discount*

In [54]:
df.discount

0       0.17
1       0.13
2       0.22
3       0.08
4       0.13
        ... 
1045    0.03
1046    0.11
1047    0.08
1048    0.03
1049    0.25
Name: discount, Length: 1050, dtype: object

The discount column contains a mix of numeric values like 0.12 and 0.1, along with some nulls and "-" as placeholders. Since it's currently stored as an object type (string), which is not appropriate for numeric operations, we'll convert it to a proper numeric type (float).

In [55]:
# Convert discount to float
df['discount'] = pd.to_numeric(df['discount'], errors='coerce')

*revenue*

In [37]:
df.revenue

0       $6,631.27
1         1430.75
2        10215.72
3        14802.98
4          2296.1
          ...    
1045     10922.39
1046     31138.37
1047      1688.18
1048      4458.65
1049      7020.29
Name: revenue, Length: 1050, dtype: object

The revenue column is currently of type object and contains values like "$6,631.27", which can't be directly converted to numeric due to the dollar sign and comma.

So before converting it to a numeric type, we’ll first remove the leading dollar sign  and comma.

In [38]:
# Remove the leading dollar sign from the 'revenue' values
df['revenue'] = df['revenue'].str.lstrip('$')

# Remove commas from the 'revenue' values to prepare for numeric conversion
df['revenue'] = df['revenue'].str.replace(',', '', regex=False)

In [39]:
# Convert to numeric type
df['revenue'] = pd.to_numeric(df['revenue'], errors='coerce')

*priority*

In [40]:
df['priority'].unique()

array(['medium', 'HIGH', 'High', 'Medium', nan, 'Low', 'high', 'low'],
      dtype=object)

In the priority column, we have values with inconsistent capitalization.
To make the values consistent, we'll standardize the text format, so that all values follow the same casing (e.g., 'High', 'Medium', 'Low').

In [61]:
df.priority=df.priority.str.capitalize()

Furthermore, we can convert the priority column into a pandas Categorical data type with a defined order. This helps pandas understand the logical sequence of priorities (i.e., 'Low' < 'Medium' < 'High') for sorting and comparisons.

In [62]:
priority_order = ['Low', 'Medium', 'High']

df['priority'] = pd.Categorical(
    df['priority'],
    categories=priority_order,
    ordered=True
)

In [66]:
df.priority.sort_values()

1049    Low
269     Low
706     Low
266     Low
265     Low
       ... 
1017    NaN
1018    NaN
1025    NaN
1032    NaN
1038    NaN
Name: priority, Length: 1050, dtype: category
Categories (3, object): ['Low' < 'Medium' < 'High']

*sales_rep_email*

In [77]:
# Convert all email addresses in 'sales_rep_email' column to lowercase
df['sales_rep_email'] = df.sales_rep_email.str.lower()

In [81]:
# Returns rows where sales_rep_email is NULL
df[df['sales_rep_email'].isnull()]

order_id     customer_name   region               product   sales_rep  \
4     ELX-1004          Techmart     East      Wireless Charger  Mike Davis   
5     ELX-1005         Gadgethub    North          Phone Holder  Bob Wilson   
17    ELX-1017     Digital World     East  Apple Watch Series 9  John Smith   
23    ELX-1023       Smart Store  Central        Gaming Headset    Jane Doe   
32    ELX-1032       Smart Store    South  Samsung  Galaxy  S24  Mike Davis   
...        ...               ...      ...                   ...         ...   
1026  ELX-1775          Techmart     East           USB-C Cable    Jane Doe   
1032  ELX-1832  Electronics Plus    North       Fitbit Charge 6  Bob Wilson   
1036  ELX-1021  Electronics Plus  Central  Apple Watch Series 9  John Smith   
1039  ELX-1014          Techmart  Central       Fitbit Charge 6  Bob Wilson   
1042  ELX-1153     Digital World    North           USB-C Cable  John Smith   

     order_date  quantity  unit_price  discount  shipping_cost     revenue  \
4    2023-05-16         3      879.73      0.13          24.46      2296.1   
5    2023-11-18        40     1064.26      0.09           8.85  $38,739.06   
17   2023-04-05        37     1137.14      0.02          14.72     41232.7   
23   2023-08-31        27      240.64      0.02          13.62     6367.33   
32   2023-02-11        27      793.99      0.14          11.33    18436.45   
...         ...       ...         ...       ...            ...         ...   
1026 2023-08-07        24      472.54      0.00          11.90    11340.96   
1032 2023-12-04        36      223.54      0.21          24.28     6357.48   
1036 2023-02-06        46      762.86      0.06          23.11    32986.07   
1039 2023-02-27        19      530.20      0.11          24.08     8965.68   
1042 2023-08-12         2      559.81      0.02          11.26     1097.23   

      total_cost priority sales_rep_email  
4        2320.56   Medium             NaN  
5       38747.91      NaN             NaN  
17      41247.42      Low             NaN  
23       6380.95      Low             NaN  
32      18447.78   Medium             NaN  
...          ...      ...             ...  
1026    11352.86     High             NaN  
1032     6381.76      NaN             NaN  
1036    33009.18     High             NaN  
1039     8989.76     High             NaN  
1042     1108.49      Low             NaN  

[151 rows x 14 columns]